### ISSUE TITLE
Mahindra (LSP) - compliance check

### ISSUE THREAD

**Issue by abhinavv245** [ONDC-Official/verification-logs#1081]
## Flow 1

- Incorrect Flow as per context/timestamps - (search,init,confirm,on_confirm,on_search,on_init,on_status)
- context/timestamp difference between /on_action and /action APIs should not be more than 1 sec

**/on_search**

- Two categories are not required for the same TAT, Immediate delviery should be used for deliveries under 60 min
- name and short_desc should include 'RTO quote' for RTO Items I3,I6
- RTO quote items linked to forward shipment quote through parent_item_id (cannot be empty string)

**/on_confirm**

- context must have required property 'bap_uri'
- message/order/fulfillments/agent empty objects should be removed
- provider location to be provided only if provided in /on_search ( to be used for cases where shipment has to be dropped off at LSP location; not required for "Immediate Delivery)
- message/order/billing must have required property 'phone'
- message/order/billing must have required property 'tax_number'
- message/order/billing must have required property 'created_at'
- message/order/billing/name mismatches in /billing in /confirm and /on_confirm
- message/order/billing/updated_at mismatches in /billing in /confirm and /on_confirm

**/on_status**

- context must have required property 'city'
- message/order/billing must have required property 'phone'
- message/order/billing must have required property 'tax_number'
- message/order/billing/address must have required property 'name'
- /on_status call for Fulfillment state - 'Order-picked-up' missing

## Flow 2

- similar issues as Flow 1

**/on_cancel**

- context must have required property 'city'

**/on_update**

- start/time/range is required in /fulfillments when ready_to_ship = yes in /update
- end/time/range is required in /fulfillments when ready_to_ship = yes in /update
- context must have required property 'city'"
- Message ID should be same as /update: \"0ad61921-7178-40da-9d34-8af4ac08dde4\""
- order must have required property 'items'" (refer API contract)
- order/fulfillments/agent empty object should be removed

**/on_status**

- /on_status call for Fulfillment state - 'Order-picked-up' missing
- provider location to be provided only if provided in /on_search ( to be used for cases where shipment has to be dropped off at LSP location; not required for "Immediate Delivery)
- Delivery timestamp (fulfillments/end/time/timestamp) cannot be provided for fulfillment state - Out-for-delivery
- RTO Pickup (fulfillments/start/time/timestamp) time cannot be future dated for fulfillment state  - RTO-Initiated
- RTO Delivery timestamp (fulfillments/end/time/timestamp) is missing for fulfillment state - RTO-Delivered
- RTO Pickup time (fulfillments/start/time/timestamp) cannot change for fulfillment state - RTO-Delivered

@jeevansanju

**Comment by jeevansanju** [comment]
submitted the revised logs please review them

**Comment by abhinavv245** [comment]
## Flow 1

**/on_search**

- message/catalog/bpp/providers/items/time/duration is not in valid format ("PT2Hr")

**/on_confirm**

- /message/order/updated_at does not match context/timestamp - "2023-08-31T11:41:30.242Z"

**/on_update**

- context must have required property 'message_id'

**/on_status**

- Pickup timestamp (fulfillments/start/time/timestamp) cannot be future dated for fulfillment state - Order-picked-up
- Delivery timestamp (fulfillments/end/time/timestamp) cannot be provided for fulfillment state - Order-picked-up
- Pickup timestamp (fulfillments/start/time/timestamp) cannot change for fulfillment state - Out-for-delivery
- Delivery timestamp (fulfillments/end/time/timestamp) cannot be provided for fulfillment state - Out-for-delivery
- Pickup timestamp (fulfillments/start/time/timestamp) cannot change for fulfillment state - Order-delivere

## Flow 2

- similar issues as Flow 1

**/on_status** 

- order must have required property 'tags' with cancellation_reason_id after order has been cancelled.
- Pickup timestamp (fulfillments/start/time/timestamp) cannot change for fulfillment state - Cancelled
- fulfillments/start/time/range is not in sync for fulfillment state - RTO-Initiated
- RTO Pickup time (fulfillments/start/time/timestamp) cannot change for fulfillment state - RTO-Delivered
- RTO Delivery timestamp (fulfillments/end/time/timestamp) is missing for fulfillment state - RTO-Delivered

@jeevansanju

**Comment by jeevansanju** [comment]
submitted the revised logs please review them

**Comment by abhinavv245** [comment]
## Flow 1

**/on_search**

- TAT for 'Same day delivery' should be greater than 60 min (else 'Immediate delivery' should be used)

**/on_confirm**

- pickup and delivery time ranges breach the 9 min delivery TAT provided in /on_search

**/update and on_update**

- what is the purpose of this /update call as the order was marked ready_to_ship in /confirm

**/on_status**

- payment/collected_by mismatch in /confirm and /on_status (BAP)
- payment/type mismatch in /confirm and /on_status (ON-ORDER)

## Flow 2

- similar issues as Flow 1

**/on_status**

- order must have required property 'tags' with cancellation_reason_id after order has been cancelled.

@jeevansanju

**Comment by jeevansanju** [comment]
- In on_search im sending  same day delivery  "duration":"PT2H"  its greater than 60 mins right?

- can i remove update and on_update.json from both flows?

**Comment by abhinavv245** [comment]
## Flow 1

**/on_confirm**

- order/state should be 'Accepted' if auto-RTS is supported by LSP

## Flow 2

**/on_update**

- context/timestamp cannot be less than /update context/timestamp (flow incorrect)

@jeevansanju

**Comment by jeevansanju** [comment]
Hi @abhinavv245, I've submitted the revised logs can you review them.

**Comment by bluecypher** [comment]
## Common Issues

**/on_search**

- 9 min delivery doesn't seem feasible as per the pickup and drop location

**/on_confirm**

- @ondc/org/awb_no is not required for P2P fulfillments
- vehicle and agent info are only required in case of last mile delivery, after agent is assigned

**/on_status (Order-picked-up)**

- pickup timestamp is not captured correctly (can't be the same as order/created_at timestamp)

## Flow 2

**/on_update**

- order/updated_at can't be the same as order/created_at timestamp


@jeevansanju

**Comment by jeevansanju** [comment]
- "9 min delivery doesn't seem feasible as per the pickup and drop location": This TAT for the given pick up and drop was given by Map My India API without traffic
- "vehicle and agent info are only required in case of last mile delivery, after agent is assigned": In confirm API, RTS will be true and delivery agent will be assigned automatically. 
Can you kindly elaborate on the error mentioned by you for this point.

**Comment by abhinavv245** [comment]
## Flow 1 &  2 

**/on_update**

- start/time/range is required in /fulfillments when ready_to_ship = yes in /update
- end/time/range is required in /fulfillments when ready_to_ship = yes in /update

**/on_status**

- context/timestamp and RTO delivery timestamp are not valid (fulfillment state -RTO - Delivered)
- Pickup timestamp (fulfillments/start/time/timestamp) cannot be provided for fulfillment state - Agent-assigned 

@jeevansanju

**Comment by bluecypher** [comment]
## Flow 1

**/on_confirm**
- estimated pickup and delivery time ranges are mandatory when "@ondc/org/order_ready_to_ship"="yes" in /confirm

## Flow 2
- Is "RTO-Disposed" state supported?


@jeevansanju

**Comment by jeevansanju** [comment]
Yes, RTO Disposed is supported. As told by your team, we are using RTO Delivered for grocery and RTO Disposed for F&B. As the current logs submitted were for groceries, RTO Disposed couldn't be used.

**Comment by abhinavv245** [comment]
## Flow 1 and 2

**/search**

- Reverse Geocoding for `fulfillment/end` failed. Area Code 560008 not matching with 12.977203,77.631812 Lat-Long pair. (how is LSP handling this)

**/on_status**

- order/id mismatch in /confirm and /on_status

@jeevansanju

**Comment by jeevansanju** [comment]
/search
- As discussed over call with Sandeep, in such cases of mismatch in address and pincodes, delivery associate will be cancelling the order, if he is unable to find the address

**Comment by bluecypher** [comment]
@BLR-0118, The Mahindra LSP (v1.1.0) logs seem fine. Please proceed with your review.

@jeevansanju
